### PR TITLE
Dawn dev branch

### PR DESCRIPTION
**Pull request by dawndrain**
Made a couple fixes:
1. Made encoder self-attn non-causal
2. Fixed param swap in chunk iterator
3. Fixed dummy data in README

### CODE REVIEW

**Comment by lucidrains** [comment]
@dawndrain Oh hey Dawn! Did you manage to get Marge working? I've heard people are having trouble reproducing this

**Comment by dawndrain** [comment]
Hi Phil,

I got it working with the dummy data. The loss is basically random though
of course. I haven’t tried with real data yet. I want to do a very simple
experiment comparing retrieving four documents vs. zero. Right now I’m
working on integrating marge with Fairseq, since that’s the platform my
team uses at work. It would definitely be easier to do that initial
experiment in your framework instead of Fairseq though...

Best,
Dawn

**Comment by lucidrains** [comment]
@dawndrain Ok Dawn! Keep me updated! Someone has told me a bunch of googlers were having trouble replicating this, and it may come down to the way the data is sharded in the beginning. But if you and your team manage to figure it out, do let me know!

**Comment by dawndrain** [comment]
Ok, I have a couple updates.

I’ve been experimenting with enwiki8 instead of dummy data and with
retrieving 0, 1, or 4 documents.   It looks like the loss when retrieving
four documents is epsilon better than retrieving 1 (like .3% better) albeit
much slower, and it looks like retrieving 1 is no better than retrieving 0
:/.

Underwhelming results aside, I think the model that retrieves four is much,
much better at retrieval than the model that retrieves one. The model that
retrieves one just retrieves a dummy document every time (I looked at five
examples and it was always one of the same two, and always irrelevant). The
model that retrieves four is much more on point :) e.g. retrieving four
examples of users when the tgt is a user, or retrieving C programming
language when the tgt is Ada programming language.

Best,
Dawn

**Comment by dawndrain** [comment]
Eh, I think I spoke too soon. The model that retrieves four is still pretty
bad at retrieving, and the model that retrieves one is getting plenty
diverse results (I’m also not sure out it would even be possible to
consistently retrieve the same nearest neighbor...) I’m still early into
training. Hopefully the retrieval quality will improve a lot with a
stronger model.

**Comment by lucidrains** [comment]
@dawndrain thank you for relaying!

**Comment by lucidrains** [comment]
@dawndrain https://github.com/lucidrains/distilled-retriever-pytorch

**Comment by dawndrain** [comment]
Cool paper with strong results i.e. ~5% improvement over fusion-in-decoder
when matching the retriever and reader inits (DPR and T5-base/large).

One thing I like is this quick preliminary experiment they did, "Before we
proceed, let us consider the following simple experiment, which is a first
indication thatreader attention scores are indeed a strong relevance signal.
Given a question and 100 passagesretrieved with DPR, our goal is to select
the 10 best passages. When using the top 10 passages fromDPR instead of the
top 100, the performance of our reader drops from 48.2 EM to 42.9 EM. Onthe
other hand, if we select the top 10 documents according tothe attention
scores, the performanceonly drops to 46.8 EM."

I'm kind of baffled why they didn't do more iterations. Performance was
still increasing pretty steeply when they stopped.
[image: Screen Shot 2020-12-16 at 3.28.03 PM.png]

You're right that this sounds *a lot* like Marge. Like almost exactly the
same. So in Marge you get a measure of the cross-attn score for each
retrieved doc, and bias these scores according to the retrieval scores, the
idea being that the retriever should learn to fade or highlight documents
in proportion to their relevance, and reuse this relevance score during
retrieval. In this new paper you again get a measure of the cross-attn
score for each retrieved doc, but you don't bias the cross-attn scores
using the retrieval scores. Instead you train the retriever to produce
scores that *mimic* the average cross-attn score.

I think I prefer this new approach to marge. One thing that still bugs me
about marge is that the retriever does a bad job of communicating the
scores to the reader, and I'd expect it to do a bad job at factorizing the
cross-doc cross-attn scores. I feel like biasing attn scores could just
confuse the reader, and it seems especially unnecessary since the reader is
normally bigger than the retriever. Maybe something as simple as
additionally prepending the retrieval score before each doc would be an
improvement, although that is a discrete operation, eh. It also bugs me
that the marge objective isn't technically end-to-end, since the retriever
isn't backprop'ing directly on the loss. I really want to see an end-to-end
baseline that retrieves k (k=2 say) docs for each query, and does k
separate forward passes, and uses those relative scores to train the
retriever.

Since we want to reuse all the faiss and indexing stuff, it seems sensible
to add this loss term to the marge repo. Possibly we could combine the two
loss terms, but I feel like that could lead to something degenerate,
haven't thought it through.

It's sad that these models are still kind of dependent on the retriever
init T.T. I'm not sure how big a deal that is, but it really subtracts from
the promise of training semi-supervised end-to-end...

Best,
Dawn
